### DISCUSSION TITLE
fighting-with-nullable-reference-types-in-razor-pages/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#71] [Announcements]
# fighting-with-nullable-reference-types-in-razor-pages/

In this post I discuss C#'s nullable reference types in the context of Razor Pages, why they  don't play nicely together and how to improve the experience.

https://andrewlock.net/fighting-with-nullable-reference-types-in-razor-pages/

**Comment by thomaslevesque** [comment]
Nice article, as usual!

I also ran into the issues you describe, and I think they highlight a fundamental issue with Razor Pages: having properties that are set for some handlers, but not for others, is a big code smell IMO. It means the class does too many things. There's no reason for `Input` to be in scope in `OnGet`, because it doesn't need it; it shouldn't be a property of the model, but a parameter of the `OnPost` method (not sure if this is possible with Razor Pages). Sure, it's convenient to just have everything stuffed together in the same class... but it's wrong, and it's the reason why these nullability problems occur. I always felt a little uncomfortable with Razor Pages, and your article finally made me it realize why...

Regarding `#nullable disable`:

> This may miss nullability bugs when interacting with the rest of your application.

You can do `#nullable disable annotations`, this will disable nullability annotations in this file, but not nullability warnings when calling other parts of the code that do have the annotations.

**Reply by andrewlock** [reply]
Interesting, sounds like we have quite different feelings about Razor Pages - I actually really like them, and the issues I describe here makes me more uncomfortable with nullable reference types than the other way around :smile:

>There's no reason for Input to be in scope in `OnGet`, because it doesn't need it; it shouldn't be a property of the model, but a parameter of the `OnPost` method (not sure if this is possible with Razor Pages).

Yep, this is totally possible with Razor Pages. You can pass arguments to the handlers and they're bound as you would expect, the same as MVC parameters. I do this commonly for `{id}` route parameters for example.

The problem with this, is it complicates one of the strong points of Razor Pages - working with forms. If you pass the data in as a method handler, then you have to pass it back through on the model somewhere. Which I think leaves you in the same position as if you're using MVC. 

I think one option I didn't describe in this post, which would be closer to the MVC equivalent, would be something like this:

```csharp
public class IndexModel : PageModel
{
    [BindProperty]
    public required InputModel Input { get; set; } = new(); // <- Never null
    // ...
}
```

This is closer to what you might do with MVC I think given that the `InputModel` is used for form binding.

The problem is, I have never run into nullability issues due to a `null` `InputModel` object. Maybe that's because the pattern in Razor Pages is consistent everywhere, maybe I'm just lucky. So all these extras to make nullable reference types feel like noise with no benefit for me in Razor Pages (as opposed to elsewhere, where I definitely see the benefit)

**Reply by andrewlock** [reply]
Oh, and I didn't know about `'disable nullable annotations` - thanks! :smile: That actually solves most of my gripes, so maybe that will be my go to :wink:

**Comment by StilgarNaib** [comment]
I think Input = null! is the most appropriate solution here. It explicitly says that you are lying to the compiler because you are. You know required will be skipped by the framework so you'd better make the lie explicit. That required keyword will never serve its actual purpose - raising an error if the property is not set. For the properties I'd go with  = null! or required too rather than the ! operator because if you follow the pattern you are never using them outside if(ModelState.IsValid) and it is more important to inform the reader which properties can and which cannot be null after validation (BTW I think you can remove the Required property if you use nullable reference types, at least with MVC you can, you still need it if you want to customize the error message though). If you make all properties not nullable you not only sprinkle ! on multiple places in the code but also effectively disable the validation where it is needed  - after the model state check.

It would be cool if C# got some way to say via a property that if a method returns true certain properties somewhere inside the class are not null like we have attributes that inform the compiler that the method does a null check on its arguments. Other than that I feel like most of the nullability pain is either unavoidable (more safety means declaring more things explicitly) or in the vast majority of cases that the frameworks we use today are not designed with nullability in mind. Hopefully future frameworks will be. For example entity framework uses public setters but an alternative could be designed to use immutable records with constructors and mutation can be done with the "with" keyword. Arguably it better represents the interaction with the database as the database row is not changed until you execute SaveChanges. I believe for pretty much every piece of the stack we could have a nullability-friendly solution, the question is when it is worth replacing the current solution, when the current solution can be fixed and when it is not worth the trouble to replace a solution that cannot be fixed.

BTW your new comments plugin wants to read my private GitHub items and act on my behalf so I decided to register an account for that purpose instead. Hopefully I don't forget I have it next time

**Reply by andrewlock** [reply]
> I think Input = null! is the most appropriate solution here...That required keyword will never serve its actual purpose

You know what, I think you've convinced me :smile: This is probably the right option (despite the ugliness). Either that or setting `#disable nullable annotations` (which I haven't tried yet)

>Other than that I feel like most of the nullability pain is either unavoidable..

Yeah. Honestly, I'm not that worried about the null checking being perfect here. I mostly just want it to stop shouting at me :joy:. If I could disable the null checks for all Razor Pages inside a folder, I think I'd be happy.

> BTW your new comments plugin wants to read my private GitHub items and act on my behalf so I decided to register an account for that purpose instead. Hopefully I don't forget I have it next time

Yeah, that's a limitation of giscus (and GitHub's overly course permissions) unfortunately :roll_eyes: You can also visit the discussion directly on GitHub (Click the "reactions" link at the top of the section) which takes you to the discussion in GitHub directly, which you can comment on safely. It's something I just haven't got around to fixing yet unfortunately (too busy working on the book!:sweat_smile:

**Reply by thomaslevesque** [reply]
> You can also visit the discussion directly on GitHub (Click the "reactions" link at the top of the section) which takes you to the discussion in GitHub directly, which you can comment on safely.

Note that the discussion is created when the first comment is posted, so this will only work if someone has already commented.

**Comment by bzamora** [comment]
Hey, great to see I am not the only who still likes razor pages 😂

The way I solved these warnings was to initialize the objects, I didn’t find any issues with razor pages like binding or validation when using that approach. I.e. 

> public InputModel Input { get; set; } = new();
> //…
> public string FirstName { get; set; } = string.Empty;

etc…

I liked that better than annotations and to me made sense as objects are initialized and therefore not null.

Use the same approach with EF as it has similar warnings for sets.

**Comment by TomSkidmore** [comment]
I think that's an interesting discussion with Blazor too. Thinking more around injected services, should these be marked as nullable also? Similar to what you mentioned, these services __should__ be not null providing you've set them up correctly (registering the services), but can be null if not.

Honestly, I think there's a fine line between keeping the compiler happy and keeping the developer happy, a trade off of not having to place `!` or `?` everywhere for a nullable use case that probably won't ever happen

**Comment by JohnGalt1717** [comment]
To me it's even worse:

1. If you follow best practices then you'll also get CA1034 warnings because public class InputModel nested inside of your PageModel
2. Using required isn't really valid because it is null in Get (or could be unless you tell it not to be)


It would be far better if InputModel was passed into OnPost(Async) and was defined like a DTO.  This would also make it far more WebApi-like.

The class should have properties that are read by the CSHTML and the CSHTML should post back as parameters on the method.

Then you can be explicit about your validation and requirements on post.

But C# also needs the "late" keyword like Dart has to say "Yes, this isn't specified here, but we will always set this before we use it. The most obvious PITA on this one that results in null! is on Unit Tests where you use the TestStartup stuff which isn't injected in the constructor. (but that one is really, we need DI for Unit Tests that is native and automatic and injects everything into the constructor.)

**Comment by jeremyregnerus** [comment]
I'm running into this right now with Blazor. I have really liked enabling nullable reference types in all of my libraries. When I enabled it in my Blazor app, I was immediately confronted with a lot of these same how's and why's. I'm especially head scratching when it comes to [Inject] private properties, as properties have to be public in order to use the the required keyword. I don't really want to make them public in this case, as one, I don't ever see a need for anything else to get them from these Blazor pages; and two, I really really don't feel like documenting these properties for who's scope should never be outside of these classes. For those cases, I'm going to stick with = null!; for now.

For objects that created from my services, I'm a bit torn. Most of my pages can't pull an object that doesn't exists (just the way this was created from the ground up), so now I'm left what to do with these before OnInitializedAsync. I make them nullable, and I'm slinging ? and ! all over my code, just to remind me to add an @if(obj is not null) on the html side.

**Comment by schwastek** [comment]
This is something I debate often.

**Should API request models allow all properties to be nullable since an empty POST request is possible?
Or should required fields use the `required` modifier?**

In other words, should we design request models based on what the API might receive (everything nullable)
or what the user is expected to provide (required fields are `required`)?

* If we send request with missing required property, we get automatic HTTP 400 response, because of **model binding error**.
    * Why model binding error? JSON deserialization error since required property is missing.
    * If automatic HTTP 400 response is disabled, the request itself becomes null.
    * It behaves like `[JsonRequired]` ([source](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/required-properties))
* If we send request with missing `null!` property, we get automatic HTTP 400 response, because of **model validation error**.
    * Why model validation error? Because non-nullable properties are implicitly attributed with `[Required(AllowEmptyStrings = true)]` ([source](https://learn.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-8.0#non-nullable-reference-types-and-required-attribute))
    * The request as a whole is not null like in the previous case. Only the missing property is.

Additionally:

* If you make mandatory properties nullable, you need to map them to non-nullable models. Otherwise, you'll need to use `!` (null-forgiving) operator a lot. For example, if you're using MediatR with validation pipeline, your query/command model is validated already, but you'll need to `!` everywhere in handler's logic.

**What's your approach nowadays?**
